Misc integration test fixes

cathay4t opened the pull request:
* Integration test: Fix python 3 split() TypeError
 * Integration test: Fix vlan test for ordering
 * Integration test: Fix python3 float TypeError
 * Integration test: Fix vlan rollback test
 * Integration test: Fix bond rollback test

EdDev commented on tests/integration/testlib/cmd.py:
```diff
@@ -41,7 +41,7 @@ def exec_cmd(cmd, env=None):
     logging.debug(_retcode_log_line(p.returncode, err=err))
 
-    return p.returncode, out, err
+    return p.returncode, out.decode('utf-8'), err.decode('utf-8')
```
In production code I would have asked to create a wrapper to convert bytes to unicode, but for tests it should be fine.
(reasoning that the caller may actually want to process bytes)

Anyway, it will be nice if you add a word or two in the doctext to mention that the output is unicode.

EdDev commented on tests/integration/nmstatectl_show_test.py:
```diff
@@ -27,7 +27,7 @@
 RC_FAIL2 = 2
 
-LOOPBACK_JSON_CONFIG = b"""        {
+LOOPBACK_JSON_CONFIG = """        {
```
How come `u` is not needed for py2? The `decode` returns unicode.

cathay4t replied:
No idea who python2 handing it with out failure. But I will convert the return value to string by using `unicode.encode('utf-8')`.

EdDev replied:
I think you need `decode`, that should fit both py2 and py3: Converting binary/bytes to text.

EdDev commented on tests/integration/testlib/assertlib.py:
```diff
@@ -34,6 +35,8 @@ def assert_state(desired_state_data):
     full_desired_state.remove_absent_entries()
     full_desired_state.normalize()
 
+    full_desired_state.state[INTERFACES].sort(key=itemgetter('name'))
```
Things are sorted in the `normalize` method, please try to move it there.

EdDev commented on tests/integration/vlan_test.py:
```diff
@@ -79,8 +79,7 @@ def test_rollback_for_vlans(eth1_up):
 
-    current_state_after_apply = netinfo.show()
-    assert current_state == current_state_after_apply
+    assertlib.assert_state(current_state)
```
If you found things failing here it is indeed interesting.
I'm pretty sure with this change it is a different assert then intended.

The rollback is supposed to revert the changes and we want to check that the new state is identical to the previous one. If we are to touch something, it is only the ordering/sorting, `assertlib.assert_state` does much more.

cathay4t replied:
OK. I will do ordering here instead of use `assertlib.assert_state()`.

cathay4t commented:
ci test please.

cathay4t commented:
ci test please.

cathay4t commented:
V2 patch set uploaded.

Changes since V1:
 * Improved previous patch based on feedback.
 * Added two patch to fix the random failure on bond and vlan test. I guess CI should be stable now.

cathay4t commented:
> I think bytes are ‘decode’ and strings are ‘encode’. Therefore you now return bytes.
> It’s like doing ‘not’ twice, so I do not understand what is going on here.

Another difference between python 2 vs 3. In python 2 `b'abc'.decode('utf-8').encode('utf-8')` got string,
python 3 got bytes. I will find a better way. Thanks.

cathay4t commented:
V3 patch set uploaded.

Changes since V2:
 * Fix the code for converting bytes to string. Now using `str(bytes.decode('utf-8'))`.
 * Replaced the patch for vlan interface ordering. New patch will sort the interface in `netinfo.show()`.

cathay4t commented:
> > Another difference between python 2 vs 3. In python 2 b'abc'.decode('utf-8').encode('utf-8') got string,
> > python 3 got bytes. I will find a better way.
> 
> In py2 bytes == str while on py3 you got both.
> On the other hand, in py3, str is always unicode while on py2 you need to specify it (decode will do the job).
> We are interested, in both py2 and py3 to get the same type, which is unicode. Therefore, I think the correct solution is to just use `decode` and nothing else.
> (bytes.decode() in py3 == string.decode() in py2)

Done. Thanks.

EdDev commented on libnmstate/netinfo.py:
```diff
@@ -15,6 +15,7 @@
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 #
 from __future__ import absolute_import
+from operator import itemgetter
```
I never got the point of using a helper like this instead of a lambda. Feels like wasting an import line.

cathay4t replied:
https://stackoverflow.com/questions/11287207/why-should-i-use-operator-itemgetterx-instead-of-x

And pylint will complain if you use lambda instead of itemgetter.
